### ISSUE TITLE
Add support for actor reminders

### ISSUE THREAD

**Issue by sebastianburckhardt** [Azure/azure-functions-durable-extension#716]
When writing samples it became clear that there are situations where an actor wants to send itself a message with a time delay. 

I believe we should support this, perhaps by adding a method along the lines of

```csharp
ctx.ScheduleOperation(TimeSpan delay, string operationName, object operationContent);
```

**Comment by cgillum** [comment]
Agreed, this would be good to have.

**Comment by olitomlinson** [comment]
This is a very desirable feature of an Actor!

While your in the design phase, it might be worth checking out Service Fabric Actor Timers & Reminders for some inspiration!

https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-actors-timers-reminders

**Comment by sebastianburckhardt** [comment]
After going through a few designs my plan is to support as a more general feature which allows an optional `scheduleAtUtc` argument to all the SignalActor functions:

```csharp
ctx.SignalEntity(entityId, operationName, operationInput, scheduleAtUtc);
```

Then a "reminder" requirement can be implemented by having an actor send a signal to itself, scheduled for some time in the future, e.g.:

```csharp
ctx.SignalEntity(ctx.Self, operationName, operationInput, DateTime.UtcNow + TimeSpan.FromSeconds(30));
```

I ran into a gating issue azure/durabletask#285 so I'll probably delay work on this for a while.

**Comment by cgillum** [comment]
@sebastianburckhardt The underlying blockers have been resolved for this work item. Do you think you can resume working on it and have it in for beta3? I have a use case for it that I'd like to try out. :)

**Comment by sebastianburckhardt** [comment]
Yes, I have already resumed work on this. Currently in the debugging stage... some things are still not quite working as expected inside DTFx, when combining timers and ContinueAsNew.

**Comment by jeffhollan** [comment]
Very excited about this feature - great to hear it's unblocked 🤞

![](https://media0.giphy.com/media/rVbAzUUSUC6dO/giphy.gif)

**Comment by sebastianburckhardt** [comment]
I fixed one problem, but then hit another DTFx issue where timers cause duplicate activations of entities: azure/durabletask#318. We'll have to fix that first.

**Comment by tomkuijsten** [comment]
Would love to test this feature. @sebastianburckhardt will there also be an option to cancel / postpone the timer?

**Comment by cgillum** [comment]
Closing completed issues in preparation for the v2.1 release.

To answer @tomkuijsten's question, we don't have an ability to cancel/postpone timers in this release. Please feel free to open an issue requesting this as a new feature and let us know how you would intend to use it.

**Comment by olitomlinson** [comment]
@cgillum Is there a time limit on how far into the future a reminder can be scheduled for? (Similar to the 7 day limit on orchestration timers)

**Comment by cgillum** [comment]
@olitomlinson yes, the **6** day limit applies here as well. :)

**Comment by olitomlinson** [comment]
@cgillum thank you.

FYI This is a real problem for my use-case as I am looking to use the built-in reminder to schedule a tidy-up operation on each Entity so that it complies with GDPR policy. 

In my use-case I will accumulate hundreds of millions of Entitys that will need to redact parts of their state at completely arbitrary times ranging from a few seconds to a few years.

I will have to externalise this part of the process and utilise Service Bus to schedule the job at the right time.

**Comment by ajdvoynos** [comment]
> Would love to test this feature. @sebastianburckhardt will there also be an option to cancel / postpone the timer?

@sebastianburckhardt Is there a way to cancel the timer?
